### PR TITLE
Restrict T.absurd argument to variables

### PR DESCRIPTION
**Pull request by aprocter-stripe**
### Motivation

See #3462.

### Test plan

Added some checks to `exhaustiveness.rb` that various kinds of non-variable expressions boink as expected.

Added some "positive" tests for all the allowed cases.

### CODE REVIEW

File: cfg/builder/builder_walk.cc [hunk 1]
```diff
@@ -344,11 +344,20 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::Expression *what, BasicBlock
                                 return;
                             }
 
-                            if (ast::isa_tree<ast::Send>(s->args[0])) {
-                                // Providing a send is the most common way T.absurd is misused
+                            if (!ast::isa_tree<ast::Local>(s->args[0]) &&
+                                !ast::isa_tree<ast::UnresolvedIdent>(s->args[0])) {
```
Comment (aprocter-stripe) [inline cfg/builder/builder_walk.cc:348]: Wasn't sure from #3462 if we just want to allow just local variables, or if we also want to allow global/class/instance vars. Conceptually it seems reasonable to me to allow all four.

**Comment by jez** [comment]
test comment

File: test/testdata/infer/exhaustiveness.rb [hunk 2]
```diff
@@ -164,6 +160,74 @@ def cant_call_only_absurd(x)
...
+
+sig {void}
+def absurd_not_reached_on_global_var
+  if $some_global
+    return
+  else
+    return
+  end
+
+  T.absurd($some_global)
```
Comment (jez) [inline test/testdata/infer/exhaustiveness.rb:174]: wow i think this is the first usage of `T.absurd` like this, completely outside of an `if / else` 😅 
nice to see it just works

**Review by jez — APPROVED** [review]
Nice!

File: test/testdata/infer/exhaustiveness.rb (hunk 2 above)
Comment (aprocter-stripe) [reply test/testdata/infer/exhaustiveness.rb:174]: #TrustTheCFG

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_IEYrvv2VXbh1zh

**Pull request merged**